src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,242 @@
+                                    && attributeTypeSymbols.Any(s => methodSymbol.HasAttribute(s))
+                                    && parameters.Length == 1
+                                    && parameters[0].Type.Equals(streamingContextTypeSymbol);
+                                var flagImplementOnDeserializationMethod = methodSymbol.IsOnDeserializationImplementation(IDeserializationCallbackTypeSymbol);
+                                var flagImplementDisposeMethod = methodSymbol.IsDisposeImplementation(compilation);
+                                var flagIsFinalizer = methodSymbol.IsFinalizer();
+
+                                if (!flagHasDeserializeAttributes && !flagImplementOnDeserializationMethod && !flagImplementDisposeMethod && !flagIsFinalizer)
+                                {
+                                    continue;
+                                }
+
+                                FindCalledDangerousMethod(methodSymbol, visited, results);
+
+                                foreach (var result in results[methodSymbol])
+                                {
+                                    compilationAnalysisContext.ReportDiagnostic(
+                                        methodSymbol.CreateDiagnostic(
+                                            Rule,
+                                            methodSymbol.ContainingType.MetadataName,
+                                            methodSymbol.MetadataName,
+                                            result.MetadataName));
+                                }
+                            }
+                        });
+
+                    /// <summary>
+                    /// Analyze the method to find all the dangerous method it calls.
+                    /// </summary>
+                    /// <param name="methodSymbol">The symbol of the method to be analyzed</param>
+                    /// <param name="visited">All the method has been analyzed</param>
+                    /// <param name="results">The result is organized by &lt;method to be analyzed, dangerous method it calls&gt;</param>
+                    void FindCalledDangerousMethod(IMethodSymbol methodSymbol, HashSet<IMethodSymbol> visited, Dictionary<IMethodSymbol, HashSet<IMethodSymbol>> results)
+                    {
+                        if (visited.Add(methodSymbol))
+                        {
+                            results.Add(methodSymbol, new HashSet<IMethodSymbol>());
+
+                            foreach (var child in callGraph[methodSymbol].Keys)
+                            {
+                                if (dangerousMethodSymbols.Contains(child))
+                                {
+                                    results[methodSymbol].Add(child);
+                                }
+                                else if (child.IsInSource())
+                                {
+                                    if (results.TryGetValue(child, out var result))
```

Comment:
You can simplify the else if part as follows:

```csharp
else if (child.IsInSource())
{
   FindCalledDangerousMethod(child, visited, results);
   results[methodSymbol].UnionWith(results[child]);
}
```

The fact that you check `if (visited.Add(methodSymbol))` at start of `FindCalledDangerousMethod` should ensure equivalence.